Add config-flow to NextBus

ViViDboarder opened the pull request:
## Proposed change
Switch to config flow. This is based on #92211. Those changes will be removed from the diff once that PR is merged.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/27198

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

epenet commented on homeassistant/components/nextbus/const.py:
```diff
@@ -0,0 +1,9 @@
+"""NextBus Constants."""
+DOMAIN = "nextbus"
+
+CONF_AGENCY = "agency"
+CONF_ROUTE = "route"
+CONF_STOP = "stop"
```
You should be able to move these four existing constants in a preliminary PR

epenet reviewed (changes requested):
A few things to get started.
Mostly it's about splitting the PR, as it is way too big to review right now.
You should be able to split with:
- constants in one PR
- iot class in one PR
- coordinator in one PR
- config-flow in one PR
- etc...

ViViDboarder committed:
NextBus: Use config flow

ViViDboarder commented:
I've refactored everything out and will push the coordinator change in the next PR. This ends up being pretty close to the same size since the Config Flow is a pretty significant change in and of itself.

epenet commented:
Please 🙏 leave as draft until preliminary pr are merged.

ViViDboarder commented:
@epenet would you be able to take a look at any of the dependent PRs to help get this across the finish line, or have any advice on getting there?

epenet reviewed (changes requested):
Coverage on config_flow.py needs to be 100%

```console
Name                                              Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------------
homeassistant/components/nextbus/config_flow.py     117      7    94%   95, 152, 183, 187, 229, 233, 255
```

epenet commented on homeassistant/components/nextbus/sensor.py:
```diff
@@ -34,6 +35,26 @@
+    discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Initialize nextbus import from config."""
+    _LOGGER.warning(
```
The warning should be replaced with an issue in the issue registry.
Look for `deprecated_yaml` in the code base.

epenet commented on homeassistant/components/nextbus/sensor.py:
```diff
@@ -34,6 +35,26 @@
+async def async_setup_platform(
+    hass: HomeAssistant,
+    config: ConfigType,
+    add_entities: AddEntitiesCallback,
```
```suggestion
    async_add_entities: AddEntitiesCallback,
```

epenet commented on homeassistant/components/nextbus/__init__.py:
```diff
@@ -1 +1,32 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up platforms for NextBus."""
+    entry_agency = entry.data.get(CONF_AGENCY)
+    if not entry_agency:
+        return False
```
This should not be necessary.
It should instead be checked as part of the config flow or the import.

epenet replied:
Yes, you should remove the redundant check here.
The config flow has the agency as compulsory.

ViViDboarder replied:
Got it. I'll do similar between steps in the config flow then. Since Agency is compulsory, I'll not check for an invalid value and rely on the Schema to do that for me.

epenet commented on homeassistant/components/nextbus/sensor.py:
```diff
@@ -67,26 +88,27 @@ def validate_tags(client, agency, route, stop):
-    stop = config[CONF_STOP]
-    name = config.get(CONF_NAME)
-
     client = NextBusClient(output_format="json")
```
This should probably be moved to `__init__.py`, and made available inside `hass["data"]`.
Right now you have only one platform but maybe in the future you will have more.

ViViDboarder replied:
The client has been moved in the Data Coordinator PR #92218

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,271 @@
+
+        await self.hass.async_add_executor_job(job)
+
+    async def async_step_import(
+        self, config_input: UserConfig | None = None
+    ) -> FlowResult:
+        """Handle import of config."""
+        if not config_input:
+            return self.async_abort(
+                reason="Import failed due to no data to migrate",
+            )
```
When would this be `None`? Surely the `import_data` is never `None`.

ViViDboarder replied:
It shouldn't be, but I added checks for edge cases like this because the method input does account for `None` as well.

epenet replied:
No it shouldn't.
```suggestion
    async def async_step_import(
        self, config_input: dict[str, Any]
    ) -> FlowResult:
        """Handle import of config."""
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,271 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+def _agency_schema(agency_names: Iterable[str]) -> vol.Schema:
+    return vol.Schema({vol.Required(CONF_AGENCY_NAME): vol.In(sorted(agency_names))})
```
Please look at `SelectSelector` and `SelectSelectorConfig` instead.
Also, there is no need to create a function that is only used once.

ViViDboarder replied:
I see some examples in the code, but no documentation. The only thing that comes up on the documentation site is something about localization. https://developers.home-assistant.io/docs/internationalization/core?_highlight=selectselectorconfig#selectors

I'll dig through the code and try to replicate.

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,271 @@
+    return vol.Schema({vol.Required(CONF_AGENCY_NAME): vol.In(sorted(agency_names))})
+
+
+def _route_schema(route_names: Iterable[str]) -> vol.Schema:
+    return vol.Schema({vol.Required(CONF_ROUTE_NAME): vol.In(sorted(route_names))})
```
Please look at `SelectSelector` and `SelectSelectorConfig` instead.
Also, there is no need to create a function that is only used once.

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,271 @@
+    return vol.Schema({vol.Required(CONF_ROUTE_NAME): vol.In(sorted(route_names))})
+
+
+def _stop_schema(stop_names: Iterable[str]) -> vol.Schema:
+    return vol.Schema({vol.Required(CONF_STOP_NAME): vol.In(sorted(stop_names))})
```
Please look at `SelectSelector` and `SelectSelectorConfig` instead.
Also, there is no need to create a function that is only used once.

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,271 @@
+    return vol.Schema({vol.Required(CONF_STOP_NAME): vol.In(sorted(stop_names))})
+
+
+def _name_schema(default: str) -> vol.Schema:
+    return vol.Schema({vol.Optional(CONF_NAME, default=default): str})
```
`CONF_NAME` should not be part of the end schema.
Instead:
- on import, you should set the title to be the original name
- on manual, you should set a good default that the user can then change

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,271 @@
+            errors=errors,
+        )
+
+    async def async_step_name(
+        self,
+        user_input: UserConfig | None = None,
+    ) -> FlowResult:
+        """Set name."""
+        errors: dict[str, str] = {}
```
This whole function should be dropped. Don't ask for the name.

ViViDboarder commented:
As of a couple days ago, the API seems to be returning no prediction results. I'm unsure if they are deprecating this API or not. I reached out a few years ago over email, but didn't get any response.

I'd still like to see this get merged and then try to find another API in the case that the lack of results is temporary.

ViViDboarder commented:
Oh, good. The API is working again.

ViViDboarder commented:
Ok, got coverage now up to 100%. I'm not sure how Codecov counts loc, but my ~1k line patch had 4 lines without coverage and was somehow under 97%.

I'm glad I worked it out through those last couple lines because a squashed a couple bugs too.

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,264 @@
+    """Handle Nextbus configuration."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
```
Not needed
```suggestion
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,264 @@
+        self._stop_tags: dict[str, str] = {}
+        _LOGGER.info("Init new config flow")
+
+    async def _update_agency_config_options(self) -> None:
+        def job():
+            self._agency_tags = {
+                a["tag"]: a["title"] for a in self._client.get_agency_list()["agency"]
+            }
+
+        await self.hass.async_add_executor_job(job)
```
I would make this a simple non async helper function outside the config flow.
Then I would call it when needed inside the config flow, wrapped with `async_add_executor_job`.

```python
def _get_agency_tags(client: NextBusClient) -> dict[str, str]:
    """Get agency tags."""
    return {
        a["tag"]: a["title"] for a in self._client.get_agency_list()["agency"]
    }
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,264 @@
+
+        await self.hass.async_add_executor_job(job)
+
+    async def _update_route_config_options(self, agency_tag: str) -> None:
+        def job():
+            self._route_tags = {
+                a["tag"]: a["title"]
+                for a in self._client.get_route_list(agency_tag)["route"]
+            }
+
+        await self.hass.async_add_executor_job(job)
```
I would make this a simple non async helper function outside the config flow.
Then I would call it when needed inside the config flow, wrapped with `async_add_executor_job`.

```python
def _get_route_tags(client: NextBusClient, agency_tag: str) -> dict[str, str]:
    """Get route tags."""
    return {
        a["tag"]: a["title"]
        for a in self._client.get_route_list(agency_tag)["route"]
    }
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,264 @@
+
+        await self.hass.async_add_executor_job(job)
+
+    async def _update_stop_config_options(
+        self, agency_tag: str, route_tag: str
+    ) -> None:
+        def job():
+            self._stop_tags = {
+                a["tag"]: a["title"]
+                for a in self._client.get_route_config(
+                    route_tag,
+                    agency_tag,
+                )[
+                    "route"
+                ]["stop"]
+            }
+
+        await self.hass.async_add_executor_job(job)
```
I would make this a simple non async helper function outside the config flow.
Then I would call it when needed inside the config flow, wrapped with `async_add_executor_job`.

```python
def _get_stop_tags(client: NextBusClient, agency_tag: str, route_tag: str) -> dict[str, str]:
    """Get stop tags."""
    return {
        a["tag"]: a["title"]
        for a in self._client.get_route_config(route_tag, agency_tag,)["route"]["stop"]
    }
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,264 @@
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    def __init__(self):
+        """Initialize NextBus config flow."""
+        self.nextbus_config: dict[str, str] = {}
+        self._client = NextBusClient(output_format="json")
+        self._agency_tags: dict[str, str] = {}
+        self._route_tags: dict[str, str] = {}
+        self._stop_tags: dict[str, str] = {}
+        _LOGGER.info("Init new config flow")
```
```suggestion
    _agency_tags: dict[str, str]
    _route_tags: dict[str, str]
    _stop_tags: dict[str, str]

    def __init__(self):
        """Initialize NextBus config flow."""
        self._client = NextBusClient(output_format="json")
        self._data: dict[str, str] = {}
        _LOGGER.debug("Init new config flow")
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,264 @@
+                reason="Import failed due to no data to migrate",
+            )
+
+        await self._update_agency_config_options()
+        self.nextbus_config[CONF_AGENCY] = config_input[CONF_AGENCY]
```
```suggestion
        agency_tags = await self.hass.async_add_executor_job(_get_agency_tags, self.client)
        if (agency := config_input[CONF_AGENCY]) not in agency_tags:
            return self.async_abort(reason="invalid_agency")
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,264 @@
+        await self._update_agency_config_options()
+        self.nextbus_config[CONF_AGENCY] = config_input[CONF_AGENCY]
+
+        await self._update_route_config_options(config_input[CONF_AGENCY])
+        self.nextbus_config[CONF_ROUTE] = config_input[CONF_ROUTE]
```
```suggestion
        route_tags = await self.hass.async_add_executor_job(_get_route_tags, self.client, agency)
        if (route := config_input[CONF_ROUTE]) not in route_tags:
            return self.async_abort(reason="invalid_route")
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,264 @@
+        await self._update_route_config_options(config_input[CONF_AGENCY])
+        self.nextbus_config[CONF_ROUTE] = config_input[CONF_ROUTE]
+
+        await self._update_stop_config_options(
+            config_input[CONF_AGENCY], config_input[CONF_ROUTE]
+        )
+        self.nextbus_config[CONF_STOP] = config_input[CONF_STOP]
```
```suggestion
        stop_tags = await self.hass.async_add_executor_job(_get_stop_tags, self.client, agency, route)
        if (stop := config_input[CONF_STOP]) not in stop_tags:
            return self.async_abort(reason="invalid_stop")
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,264 @@
+        )
+        self.nextbus_config[CONF_STOP] = config_input[CONF_STOP]
+
+        self.nextbus_config[CONF_NAME] = (
+            config_input.get(CONF_NAME)
+            or f"{config_input[CONF_AGENCY]} {config_input[CONF_ROUTE]}"
+        )
```
Not needed
```suggestion
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,264 @@
+            or f"{config_input[CONF_AGENCY]} {config_input[CONF_ROUTE]}"
+        )
+
+        errors = self._validate_config()
+        if errors:
+            _LOGGER.error(errors)
+            return await self.async_step_user()
```
Not needed
```suggestion
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,264 @@
+            _LOGGER.error(errors)
+            return await self.async_step_user()
+
+        return self.async_create_entry(
+            title=self.nextbus_config[CONF_NAME],
+            data=self.nextbus_config,
+        )
```
Better to be explicit
```suggestion
        return self.async_create_entry(
            title=config_input.get(CONF_NAME) or f"{agency_tags[agency]} {route_tags[route]} {stop_tags[stop]}",
            data={CONF_AGENCY=agency, CONF_ROUTE=route, CONF_STOP=stop},
        )
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+        if stop not in stop_tags:
+            return self.async_abort(reason="invalid_stop")
+
+        self.data = {
```
No need to store it in an instance variable. A local variable is fine here.
```suggestion
        data = {
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+        }
+
+        # Abort if duplicate entries exist
+        self._async_abort_entries_match(self.data)
```
```suggestion
        self._async_abort_entries_match(data)
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+                config_input.get(CONF_NAME)
+                or f"{config_input[CONF_AGENCY]} {config_input[CONF_ROUTE]}"
+            ),
+            data=self.data,
```
```suggestion
            data=data,
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+        return self.async_create_entry(
+            title=(
+                config_input.get(CONF_NAME)
+                or f"{config_input[CONF_AGENCY]} {config_input[CONF_ROUTE]}"
```
I think this should match the title from user flows.
```suggestion
                or f"{agency_tags[agency]} {route_tags[route]} {stop_tags[stop]}"
```

ViViDboarder replied:
It doesn’t match because I thought it better to match existing behavior since this is for migration.

epenet replied:
I see... Yes since there is no unique id it makes sense.

Anyway that makes me realize that it would be good to migrate to attributes in a preliminary PR:
- self._attr_name = ...
- self._attr_native_value = ...
- self._attr_extra_state_attributes = ...

ViViDboarder replied:
I'm a bit unclear how I should be adding the unique ID here. I got feedback in the other PR that I shouldn't use the Agency + Route + Stop, which should be unique in that if you pass those same values to another sensor, you'd end up with a duplicate sensor providing the exact same results.

If not using those values, it's not clear to me what the advantage of providing a Unique ID that is not tied to the API request being made. It would effectively allow duplicates anyway.

Regardless, is this a blocker for getting this PR shipped? There aren't that many users here (14 according to the stats page) so I don't want to make perfect the enemy of good here and spend too much time on the config flow setup (which is typically done once) when I have another PR improving the API requests which are done many times a day.

ViViDboarder replied:
@epenet is this good now that we're rebased on the merged attr changes?

edenhaus replied:
@home-assistant/core Can you please answer the question about the unique id?

ViViDboarder replied:
Thanks @edenhaus. I added a commit to use the Entry ID as the unique ID as suggested, but I still feel it would be better to use the combination of the three config values since those are indeed point to a unique stop much like lat-long point to a unique location.

edenhaus replied:
As we now have a unique_id, I suggest equally setting the title in both flows. Otherwise, it can confuse users that the old ones have no stop in the title.

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+
+    async def async_step_user(
+        self,
+        user_input: UserConfig | None = None,
```
```suggestion
        user_input: dict[str, str] | None = None,
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+
+    async def async_step_agency(
+        self,
+        user_input: UserConfig | None = None,
```
```suggestion
        user_input: dict[str, str] | None = None,
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+
+    async def async_step_route(
+        self,
+        user_input: UserConfig | None = None,
```
```suggestion
        user_input: dict[str, str] | None = None,
```

epenet commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+    ) -> FlowResult:
+        """Set name based on agency, route, and stop."""
+
+        # Abort if duplicate entries exist
+        self._async_abort_entries_match(self.data)
+
+        agency_tag = self.data[CONF_AGENCY]
+        route_tag = self.data[CONF_ROUTE]
+        stop_tag = self.data[CONF_STOP]
+
+        agency_name = self._agency_tags[agency_tag]
+        route_name = self._route_tags[route_tag]
+        stop_name = self._stop_tags[stop_tag]
+
+        return self.async_create_entry(
+            title=f"{agency_name} {route_name} {stop_name}",
+            data=self.data,
+        )
```
Please delete `async_step_name` function, and move the content inside `async_step_stop`

epenet commented on homeassistant/components/nextbus/util.py:
```diff
@@ -1,6 +1,8 @@
 """Utils for NextBus integration module."""
 from typing import Any
 
+UserConfig = dict[str, Any]
+
```
Not needed.
```suggestion
```

ViViDboarder commented:
Rebased on dev following the merger of #94084

edenhaus commented on homeassistant/components/nextbus/strings.json:
```diff
@@ -0,0 +1,46 @@
+  "title": "NextBus predictions",
+  "config": {
+    "step": {
+      "user": {
+        "data": {
+          "agency": "Metro agency",
+          "route": "Route",
+          "stop": "Stop"
+        }
+      },
```
```suggestion
```

Step user only redirects to step agency.

ViViDboarder commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,202 @@
+        """Select stop."""
+
+        if user_input is not None:
+            self.data[CONF_STOP] = user_input[CONF_STOP]
```
For readability, I left the similar assignments grouped together. Combining these to one line will not have a performance impact yet makes it less clear how the tag values are used.

I could get rid of these intermediate variables entirely as a minor optimization, but I left them so that if a KeyError is raised, it's apparent whether or not it was with `self.data` or with `self._*_tags`.

ViViDboarder commented on homeassistant/components/nextbus/sensor.py:
```diff
@@ -34,59 +36,51 @@
+        config.data.get(CONF_NAME),
+    )
+
+    async_add_entities((sensor,))
```
A number of my tests break when I make this change. I'm reluctant to change them here as I only wrote this as an interim solution for splitting patches. This method is completely rewritten in https://github.com/home-assistant/core/pull/92218 where I implement the data update coordinator to align with best practices and reduce API calls.

edenhaus replied:
What's the reason they break?

ViViDboarder replied:
They all fail with this error:

```
2023-09-07 12:06:19.884 INFO     SyncWorker_0 homeassistant.loader:loader.py:638 Loaded nextbus from homeassistant.components.nextbus
2023-09-07 12:06:19.885 INFO     MainThread homeassistant.setup:setup.py:252 Setting up nextbus
2023-09-07 12:06:19.885 INFO     MainThread homeassistant.setup:setup.py:303 Setup of domain nextbus took 0.0 seconds
2023-09-07 12:06:19.885 INFO     SyncWorker_0 homeassistant.loader:loader.py:638 Loaded sensor from homeassistant.components.sensor
2023-09-07 12:06:19.886 INFO     MainThread homeassistant.setup:setup.py:252 Setting up sensor
2023-09-07 12:06:19.886 INFO     MainThread homeassistant.setup:setup.py:303 Setup of domain sensor took 0.0 seconds
2023-09-07 12:06:19.886 DEBUG    MainThread homeassistant.core:core.py:1032 Bus:Handling <Event component_loaded[L]: component=sensor>
2023-09-07 12:06:19.886 DEBUG    MainThread homeassistant.helpers.translation:translation.py:217 Cache miss for en: nextbus
2023-09-07 12:06:19.886 INFO     MainThread homeassistant.components.sensor:entity_platform.py:306 Setting up sensor.nextbus
2023-09-07 12:06:19.887 DEBUG    MainThread homeassistant.components.nextbus.sensor:sensor.py:71 {'agency': 'sf-muni', 'route': 'F', 'stop': '5650', 'name': 'Custom Name'}
2023-09-07 12:06:19.887 DEBUG    SyncWorker_0 homeassistant.components.nextbus.sensor:sensor.py:114 sf-muni:F:5650:Predictions results: {'predictions': {'agencyTitle': 'San Francisco Muni', 'agencyTag': 'sf-muni', 'routeTitle': 'F-Market & Wharves', 'routeTag': 'F', 'stopTitle': 'Market St & 7th St', 'stopTag': '5650', 'direction': {'title': 'Outbound', 'prediction': [{'minutes': '1', 'epochTime': '1553807371000'}, {'minutes': '2', 'epochTime': '1553807372000'}, {'minutes': '3', 'epochTime': '1553807373000'}, {'minutes': '10', 'epochTime': '1553807380000'}]}}}
2023-09-07 12:06:19.887 DEBUG    SyncWorker_0 homeassistant.components.nextbus.sensor:sensor.py:114 sf-muni:F:5650:Messages: []
2023-09-07 12:06:19.887 ERROR    MainThread homeassistant.components.sensor:entity_platform.py:539 nextbus: Error on device update!
Traceback (most recent call last):
  File "/Users/ifij/workspace/vividboarder/home-assistant/homeassistant/helpers/entity_platform.py", line 537, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/Users/ifij/workspace/vividboarder/home-assistant/homeassistant/helpers/entity.py", line 786, in async_device_update
    await hass.async_add_executor_job(self.update)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/ifij/workspace/vividboarder/home-assistant/homeassistant/components/nextbus/sensor.py", line 183, in update
    self.async_write_ha_state()
  File "/Users/ifij/workspace/vividboarder/home-assistant/homeassistant/helpers/entity.py", line 586, in async_write_ha_state
    raise NoEntitySpecifiedError(
homeassistant.exceptions.NoEntitySpecifiedError: No entity id specified for entity Custom Name
```

I get this error simply by adding the `True` value.

ViViDboarder replied:
It works if I remove `self.async_write_ha_state()` from the `update()` method in the sensor. It's unclear to me when I'd need to call that.

ViViDboarder replied:
Searching on Discord, people seem to suggest calling that when using `_attr_native_value`, however the tests are working in getting state anyway. Maybe that's outdated advice.

ViViDboarder replied:
Ah. Thanks! Should be good to go now then. I verified that it works when running the application and not just in tests.

edenhaus commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,202 @@
+
+            agency_tag = self.data[CONF_AGENCY]
+            route_tag = self.data[CONF_ROUTE]
+            stop_tag = self.data[CONF_STOP]
```
Ok, I'm fine with leaving it. Reason described in the other conversation

ViViDboarder commented:
@edenhaus is there a way to reopen #92218 once this is merged? That one was dependent on this and closed due to becoming stale since this isn't merged yet.

edenhaus commented on homeassistant/components/nextbus/sensor.py:
```diff
@@ -34,59 +36,50 @@
+        DOMAIN,
+        "deprecated_yaml",
+        is_fixable=False,
+        breaks_in_ha_version="2023.7.0",
```
```suggestion
        breaks_in_ha_version="2024.4.0",
```

We have a six month deprecation period

edenhaus commented on homeassistant/components/nextbus/sensor.py:
```diff
@@ -34,59 +36,50 @@
+    discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Initialize nextbus import from config."""
+    async_create_issue(
```
Please create the issue similar to https://github.com/home-assistant/core/blob/dev/homeassistant/components/brottsplatskartan/sensor.py#L45-L58

edenhaus commented on homeassistant/components/nextbus/strings.json:
```diff
@@ -0,0 +1,39 @@
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_service%]"
+    }
+  },
+  "issues": {
+    "deprecated_yaml": {
+      "title": "The NextBus YAML configuration has been deprecated",
+      "description": "Configuring NextBus using YAML has been deprecated.\n\nYour existing YAML configuration has been imported into the UI automatically.\n\nRemove the `nextbus` YAML configuration from your configuration.yaml file and restart Home Assistant to fix this issue."
+    }
+  }
```
```suggestion
  }
```

edenhaus commented:
> @edenhaus is there a way to reopen #92218 once this is merged? That one was dependent on this and closed due to becoming stale since this isn't merged yet.

Normally I could reopen it but unfortunately, you have force-pushed or recreated the branch afterwards. So GitHub doesn't allow me to reopen it again. Please open a new PR after this one is merged :) Next time only comment to the PR to unstale it.

ViViDboarder commented:
I just rebased onto dev and pushed again. This is because the latest change to the deprecation issue template is based on recent best practices and was not functional since the last time I rebased was too long ago.

edenhaus commented on tests/components/nextbus/test_config_flow.py:
```diff
@@ -0,0 +1,182 @@
+    )
+    await hass.async_block_till_done()
+
+    assert result.get("type") == "create_entry"
```
```suggestion
    assert result.get("type") == FlowResultType.CREATE_ENTRY
```

edenhaus commented on tests/components/nextbus/test_config_flow.py:
```diff
@@ -0,0 +1,182 @@
+    )
+    await hass.async_block_till_done()
+
+    assert result.get("type") == "abort"
```
```suggestion
    assert result.get("type") == FlowResultType.ABORT
```

edenhaus commented on tests/components/nextbus/test_config_flow.py:
```diff
@@ -0,0 +1,182 @@
+        CONF_AGENCY: "sf-muni",
+        CONF_ROUTE: "F",
+        CONF_STOP: "5650",
+    }
+    data.update(override)
```
```suggestion
        **override
    }
```

edenhaus commented on tests/components/nextbus/test_config_flow.py:
```diff
@@ -0,0 +1,182 @@
+    )
+    await hass.async_block_till_done()
+
+    assert result.get("type") == "abort"
```
```suggestion
    assert result.get("type") == FlowResultType.ABORT
```

edenhaus commented on tests/components/nextbus/test_config_flow.py:
```diff
@@ -0,0 +1,182 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result.get("type") == "form"
```
```suggestion
    assert result.get("type") == FlowResultType.FORM
```

edenhaus commented on tests/components/nextbus/test_config_flow.py:
```diff
@@ -0,0 +1,182 @@
+    )
+    await hass.async_block_till_done()
+
+    assert result.get("type") == "form"
```
```suggestion
    assert result.get("type") == FlowResultType.FORM
```

edenhaus commented on tests/components/nextbus/test_config_flow.py:
```diff
@@ -0,0 +1,182 @@
+    )
+    await hass.async_block_till_done()
+
+    assert result.get("type") == "form"
```
```suggestion
    assert result.get("type") == FlowResultType.FORM
```

edenhaus commented on tests/components/nextbus/test_config_flow.py:
```diff
@@ -0,0 +1,182 @@
+    )
+    await hass.async_block_till_done()
+
+    assert result.get("type") == "create_entry"
```
```suggestion
    assert result.get("type") == FlowResultType.CREATE_ENTRY
```

edenhaus commented on tests/components/nextbus/test_util.py:
```diff
@@ -0,0 +1,33 @@
+        ([], []),
+        (None, None),
+        ("test", "test"),
+        (["test"], "test"),
```
```suggestion
        (["test"], "test"),
        (["test", "bla"], "test"),
```

edenhaus commented on tests/components/nextbus/test_util.py:
```diff
@@ -0,0 +1,33 @@
+        (["test"], "test"),
+    ),
+)
+def test_maybe_first(input: list[Any], expected: Any) -> None:
```
```suggestion
def test_maybe_first(input: list[Any] | None, expected: Any) -> None:
```

edenhaus commented on tests/components/nextbus/test_sensor.py:
```diff
@@ -48,24 +67,19 @@
-
 @pytest.fixture
-def mock_nextbus():
+def mock_nextbus() -> Generator[MagicMock, None, None]:
```
Please move fixture used in multiple files into `conftest.py`

ViViDboarder commented on tests/components/nextbus/test_config_flow.py:
```diff
@@ -0,0 +1,182 @@
+    """Create a mock py_nextbus module."""
+    with patch(
+        "homeassistant.components.nextbus.config_flow.NextBusClient"
+    ) as NextBusClient:
+        yield NextBusClient
```
I had it this way because it's a class, and classes usually are upper case, even if they are instantiated like a variable. Eg. `MyClass = namedtuple('MyClass')`. It doesn't really matter since this is localized and never used elsewhere, so I can change it anyway.

edenhaus commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,211 @@
+        }
+
+        # Abort if duplicate entries exist
+        self._async_abort_entries_match(data)
```
```suggestion
        await self.async_set_unique_id("_".join(value for _, value in data.items()))
        self._abort_if_unique_id_configured()
```

I discussed it with @frenck, and you can use `agency + route + stop` as a unique ID for this integration.

ViViDboarder replied:
Thanks! That's great news. 👍 I'll refactor that back in.

edenhaus commented on homeassistant/components/nextbus/sensor.py:
```diff
@@ -34,59 +36,54 @@
+
+    sensor = NextBusDepartureSensor(
+        client,
+        config.entry_id,
```
```suggestion
        config.unique_id,
```

edenhaus commented on homeassistant/components/nextbus/sensor.py:
```diff
@@ -103,12 +100,13 @@ class NextBusDepartureSensor(SensorEntity):
     _attr_icon = "mdi:bus"
 
-    def __init__(self, client, agency, route, stop, name=None):
+    def __init__(self, client, entry_id, agency, route, stop, name=None):
```
```suggestion
    def __init__(self, client, unique_id, agency, route, stop, name=None):
```

edenhaus commented on homeassistant/components/nextbus/sensor.py:
```diff
@@ -103,12 +100,13 @@ class NextBusDepartureSensor(SensorEntity):
         self.route = route
         self.stop = stop
         self._attr_extra_state_attributes = {}
+        self._attr_unique_id = entry_id
```
```suggestion
        self._attr_unique_id = unique_id
```

edenhaus commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,211 @@
+        if user_input is not None:
+            self.data[CONF_STOP] = user_input[CONF_STOP]
+
+            self._async_abort_entries_match(self.data)
```
Please use the same unique_id as above

ViViDboarder commented:
@edenhaus two concerns with changing the title behavior on import. 

First, as I understand it, it would change the names of existing entities, which I also believe would be more confusing than new ones using a different convention.

Second, the way the title is set for the user flow leverages the responses from the three `_get_(agency|route|stop)_tags` functions to get the user friendly names. I was asked to refactor the import function to not call those discretely to minimize executor tasks, so I no longer get the responses and instead encapsulate the validation entirely in the other function.

I could change the response to return `(str|None, (dict, dict, dict))` and use the results returned there, so it's manageable, but it isn't something I want to do unless we're absolutely settled on changing the title.

ViViDboarder commented:
Just found a functionality bug. Some stops share the same "name" as others, but differ in that they offer routes in different directions. When the select options were built, this would result in one of the stop options being overwritten by another.

The latest commit fixes this by identifying stops with duplicative titles and appending the offered direction names. I've also sorted the list alphabetically to make it easier to find stops.

edenhaus commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,231 @@
+        self._abort_if_unique_id_configured()
+
+        return self.async_create_entry(
+            title=(
+                config_input.get(CONF_NAME)
+                or f"{config_input[CONF_AGENCY]} {config_input[CONF_ROUTE]}"
+            ),
```
```suggestion
            title=f"{agency_tags[agency]} {route_tags[route]} {stop_tags[stop]}".
```

edenhaus commented on homeassistant/components/nextbus/sensor.py:
```diff
@@ -103,12 +100,13 @@ class NextBusDepartureSensor(SensorEntity):
     _attr_icon = "mdi:bus"
 
-    def __init__(self, client, agency, route, stop, name=None):
+    def __init__(self, client, unique_id, agency, route, stop, name=None):
```
```suggestion
    def __init__(self, client, unique_id, agency, route, stop, name:str):
```

edenhaus commented on homeassistant/components/nextbus/sensor.py:
```diff
@@ -103,12 +100,13 @@ class NextBusDepartureSensor(SensorEntity):
+        self._attr_unique_id = unique_id
 
         # Maybe pull a more user friendly name from the API here
         self._attr_name = f"{agency} {route}"
```
```suggestion
        self._attr_name = name
```

Please delete the following as the name is always set:
```
if name:
            self._attr_name = name
```

ViViDboarder commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,231 @@
+            CONF_AGENCY: agency,
+            CONF_ROUTE: route,
+            CONF_STOP: stop,
+        }
```
Oh, I think I see. So set `CONF_NAME` to provide a name for the entity, but name the `ConfigEntry` consistently with user added configs? There is still the issue of not having values for `route_tags`, etc. here without the previously mentioned refactor.

It wouldn't be too hard to make that refactor, but it feels like a lot more complexity for something that is going to eventually go away.

Would you like me to make this refactor?

1. Update `_validate_import` to return `tuple[str | None, tuple[dict[str, str], dict[str, str], dict[str, str]] | None]`
2. Update setting `CONF_NAME`
3. Update title to be consistent with `user` flow

ViViDboarder replied:
I made the change.

edenhaus commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,243 @@
+
+def _validate_import(
+    client: NextBusClient, agency_tag: str, route_tag: str, stop_tag: str
+) -> tuple[str, None] | tuple[None, tuple[str, str, str]]:
```
```suggestion
) -> str | tuple[str, str, str]:
    """Validate import data and return either an error as string or a tuple of (agency, route, stop)."""
```

edenhaus commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,243 @@
+    agency_tags = _get_agency_tags(client)
+    agency = agency_tags.get(agency_tag)
+    if not agency:
+        return "invalid_agency", None
```
```suggestion
        return "invalid_agency"
```

edenhaus commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,243 @@
+    route_tags = _get_route_tags(client, agency_tag)
+    route = route_tags.get(route_tag)
+    if not route:
+        return "invalid_route", None
```
```suggestion
        return "invalid_route"
```

edenhaus commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,243 @@
+    stop_tags = _get_stop_tags(client, agency_tag, route_tag)
+    stop = stop_tags.get(stop_tag)
+    if not stop:
+        return "invalid_stop", None
```
```suggestion
        return "invalid_stop"
```

edenhaus commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,243 @@
+    if not stop:
+        return "invalid_stop", None
+
+    return None, (agency, route, stop)
```
```suggestion
    return (agency, route, stop)
```

edenhaus commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,243 @@
+        route_tag = config_input[CONF_ROUTE]
+        stop_tag = config_input[CONF_STOP]
+
+        validation_error, names = await self.hass.async_add_executor_job(
```
```suggestion
        validation_result = await self.hass.async_add_executor_job(
```

edenhaus commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,243 @@
+            route_tag,
+            stop_tag,
+        )
+        if validation_error is not None:
+            return self.async_abort(reason=validation_error)
+
+        # This cast is here because when unpacking tuples, typing thinks the `names` type
+        # can be None, despite the check against `validation_error` above. This limitation
+        # does not exist if using a single tuple and accessing by index, but using names
+        # is more clear for future debugging.
+        agency_name, route_name, stop_name = typing.cast(tuple[str, str, str], names)
```
```suggestion
        if isinstance(validation_result, str):
            return self.async_abort(reason=validation_result)
```

edenhaus commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,243 @@
+        self._abort_if_unique_id_configured()
+
+        return self.async_create_entry(
+            title=f"{agency_name} {route_name} {stop_name}",
```
```suggestion
            title=" ".join(validation_result),
```

ViViDboarder commented:
Can tests be re-run? This was a flake unrelated to the actual test. There was a 404 from the Github Actions API.

```
[2023-09-14T15:17:51.451Z] ['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}
Error: Codecov: Failed to properly upload: The process '/home/runner/work/_actions/Wandalen/codecov-action/dist/codecov' failed with exit code 255
Error: Process returned exit code 1
```

ViViDboarder commented:
I squashed and rebased so I could force a re-run.

edenhaus reviewed (approved):
Thanks @ViViDboarder :+1:

ViViDboarder commented:
Thank you! I'll re-push the dependent PR now.

MartinHjelmare commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,236 @@
+        """Initialize NextBus config flow."""
+        self.data: dict[str, str] = {}
+        self._client = NextBusClient(output_format="json")
+        _LOGGER.info("Init new config flow")
```
Please remove this log.

MartinHjelmare commented on homeassistant/components/nextbus/config_flow.py:
```diff
@@ -0,0 +1,236 @@
+            stop_tag,
+        )
+        if isinstance(validation_result, str):
+            return self.async_abort(reason=validation_result)
```
We should raise a repair issue when config import fails.

MartinHjelmare commented on tests/components/nextbus/test_config_flow.py:
```diff
@@ -0,0 +1,162 @@
+    hass: HomeAssistant, mock_setup_entry: MagicMock, mock_nextbus_lists: MagicMock
+) -> None:
+    """Test config is imported and component set up."""
+    await setup.async_setup_component(hass, "persistent_notification", {})
```
We don't need to set up the persistent_notification integration anymore.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
